src/pykeen/evaluation/restricted_evaluator.py

```diff
@@ -0,0 +1,181 @@
+        Therefore, the shape of head_samples and tail_samples will be [num_eval_triples, n].
+
+        We don't use the built-in restrict_entities_to in the default evaluator since it selects the same K
+        negatives for all evaluation triples. Instead, here each triple has its own negatives.
+
+        :param validation_factory
+            Triples to be evaluated at validation/test time
+        :param all_pos
+            All positive triples, the training graph for the transductive case,
+            inductive_inference graph for the inductive case
+        :param ks:
+            The values for which to calculate hits@k. Defaults to {1,3,5,10}.
+        :param filtered:
+            Whether to use the filtered evaluation protocol. If enabled, ranking another true triple higher than the
+            currently considered one will not decrease the score.
+        :param num_negatives: dtype: int
+            Number of negative entities (per triple) to sample. Defaults to 50 to replicate [teru2020]
+        :param head_samples: shape: (num_valid_triples, n), dtype: torch.Tensor
+            Integer ids of entities to be considered negative samples against which head prediction is evaluated.
+            If None, the evaluator will generate them using the published procedure
+        :param tail_samples: shape: (num_valid_triples, n), dtype: torch.Tensor
+            Integer ids of entities to be considered negative samples against which tail prediction is evaluated.
+            If None, the evaluator will generate them using the published procedure
+
+        """
+        super().__init__(**kwargs)
+
+        if head_samples is None or tail_samples is None:
+            self.head_samples, self.tail_samples, self.negs_dict = sample_negatives(
+                valid_triples=validation_factory, all_pos=all_pos, num_samples=num_negatives
+            )
+        else:
+            self.head_samples, self.tail_samples = head_samples, tail_samples
+            self.negs_dict = generate_dict(validation_factory, head_samples, tail_samples)
+
+
+    def _update_ranks_(
+        self,
+        true_scores: torch.FloatTensor,
+        all_scores: torch.FloatTensor,
+        side: str,
+        hrt_batch: MappedTriples,
+    ) -> None:
+        """Shared code for updating the stored ranks for head/tail scores.
+
+        :param true_scores: shape: (batch_size,)
+        :param all_scores: shape: (batch_size, num_entities)
+        :param side: head/tail
+        :param hrt_batch: actual triples in a batch on which we are evaluating, needed for dict lookup
+        """
+        # land on a cpu for dictionary lookup
+        hrt_batch = hrt_batch.cpu().tolist()
+        sampled_entities = torch.stack([
+            self.negs_dict[side][tuple(triple)] for triple in hrt_batch
+        ], dim=0).to(all_scores.device)
+
+        batch_ranks = compute_rank_from_scores(
```

Comment:
Would the following work?

```python
super()._update_ranks_(
            true_scores=true_scores,
            all_scores=all_scores.gather(1, sampled_entities),
            side=side,
            hrt_batch=hrt_batch,
        )
```

Reply:
Probably not - down in the super function there is 
```
self.num_entities = all_scores.shape[1]
```

and since `all_scores` shape changed to `[bs, num_negatives]` instead of `[bs, num_all_entities]` evaluation will crash

Reply:
is the num_entities constant throught evaluation? can't this be better set in the init function?

Reply:
afaik, both Evaluator and RankBasedEvaluator do not have any init arguments specifying on which dataset they will be executed, so this assignment happens as soon as we get the scores shape

Reply:
hmm sorry this was something that was already present in the original evaluator, so it's not your problem 🐞